src/MongoDB.Driver.Core/Core/Authentication/AwsSignatureVersion4.cs

```diff
@@ -0,0 +1,203 @@
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// The AWS signature version 4.
+    /// </summary>
+    public class AwsSignatureVersion4
+    {
+        /// <summary>
+        /// Gets canonical headers.
+        /// </summary>
+        /// <param name="requestHeaders">The request headers.</param>
+        /// <returns>The canonical headers.</returns>
+        public static string GetCanonicalHeaders(SortedDictionary<string, string> requestHeaders)
+        {
+            return string.Join("\n", requestHeaders.Select(x => $"{x.Key.ToLowerInvariant()}:{x.Value}"));
+        }
+
+        /// <summary>
+        /// Gets Amazon region from host.
+        /// </summary>
+        /// <param name="host">The host.</param>
+        /// <returns>The region.</returns>
+        public static string GetRegion(string host)
+        {
+            if (host == "sts.amazonaws.com")
+            {
+                return "us-east-1";
+            }
+
+            var split = host.Split('.');
+            if (split.Count() > 1)
+            {
+                return split[1];
+            }
+
+            return "us-east-1";
+        }
+
+        /// <summary>
+        /// Gets signed headers.
+        /// </summary>
+        /// <param name="requestHeaders">The request headers.</param>
+        /// <returns>The signed headers.</returns>
+        public static string GetSignedHeaders(SortedDictionary<string, string> requestHeaders)
+        {
+            return string.Join(";", requestHeaders.Keys.Select(x => x.ToLowerInvariant()));
+        }
+
+        /// <summary>
+        /// Sign request.
+        /// </summary>
+        /// <param name="accessKey">The access key.</param>
+        /// <param name="secretKey">The secret key.</param>
+        /// <param name="securityToken">The security token.</param>
+        /// <param name="salt">The salt.</param>
+        /// <param name="host">The host.</param>
+        /// <returns>The signed request.</returns>
+        public static Tuple<string, string> SignRequest(
+            string accessKey,
+            string secretKey,
+            string securityToken,
+            byte[] salt,
+            string host)
+        {
+            return SignRequest(DateTime.UtcNow, accessKey, secretKey, securityToken, salt, host);
+        }
+
+        /// <summary>
+        /// Signs the request.
```

Comment:
For composing signature details refer to specification: [link](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#conversation-5)
Specifically this image (it is included in specification): [link](https://github.com/mongodb/specifications/raw/master/source/auth/includes/calculating_a_signature.png)